indico/modules/events/registration/client/js/form/fields/TimetableSessions.jsx

```diff
@@ -0,0 +1,257 @@
+
+function TimetableSessionsComponent({
+  value,
+  name,
+  sessionData,
+  onChange,
+  allowFullDays,
+  display,
+  minimum,
+  maximum,
+  ...props
+}) {
+  function handleValueChange(e, data) {
+    const id = data.value;
+    const newValue = value.includes(id) ? value.filter(el => el !== id) : [...value, id];
+    onChange(newValue);
+  }
+
+  const minMsg = minimum > 0 ? `at least ${minimum}` : ``;
+  const maxMsg = maximum > 0 ? `no more than ${maximum}` : ``;
+  const and = minimum > 0 && maximum > 0 ? ` and ` : ``;
+  const validatorMsg =
+    maximum > 0 || minimum > 0 ? `Please select ${maxMsg} ${and} ${minMsg} session blocks. ` : ``;
+  const currentSelectionMsg = `You selected ${value.length} session blocks.`;
+
+  return (
+    <Form.Group styleName="timetablesessions-field">
+      <strong>
+        {validatorMsg}
+        {currentSelectionMsg}
+      </strong>
+      {sessionData &&
+        Object.keys(sessionData)?.map((date, index) => (
+          <SessionBlockHeader
+            value={value}
+            index={index}
+            name={name}
+            data={sessionData[date]}
+            label={date}
+            allowFullDays={allowFullDays}
+            display={display}
+            key={`session-block-${index}`}
+            handleValueChange={handleValueChange}
+            onChange={onChange}
+            props={props}
+          />
+        ))}
+    </Form.Group>
+  );
+}
+
+TimetableSessionsComponent.propTypes = {
+  value: PropTypes.arrayOf(PropTypes.number).isRequired,
+  name: PropTypes.string.isRequired,
+  sessionData: PropTypes.object,
+  onChange: PropTypes.func.isRequired,
+  allowFullDays: PropTypes.bool.isRequired,
+  display: PropTypes.string.isRequired,
+  maximum: PropTypes.number.isRequired,
+  minimum: PropTypes.number.isRequired,
+};
+
+TimetableSessionsComponent.defaultProps = {
+  sessionData: {},
+};
+
+function SessionBlockHeader({
+  value,
+  data,
+  label,
+  allowFullDays,
+  display,
+  handleValueChange,
+  onChange,
+}) {
+  const [isChecked, setIsChecked] = useState(false);
+  const [isIndeterminate, setIsIndeterminate] = useState(false);
+  const [isExpanded, setIsExpanded] = useState(display === 'expanded');
+
+  const arrowLabel = isExpanded ? '▼' : '▶';
+  const isExpandedClass = isExpanded ? '' : 'hidden';
+  const childrenIds = data.map(e => e.id);
+
+  function handleHeaderClick() {
+    setIsExpanded(!isExpanded);
+  }
+
+  function handleHeaderChange() {
+    const newValue =
+      isChecked || isIndeterminate
+        ? value.filter(element => !childrenIds.includes(element))
+        : [...value, ...childrenIds];
+    onChange(newValue);
+  }
+
+  useEffect(() => {
+    setIsChecked(() =>
+      childrenIds.reduce((acum, id) => {
+        return acum && value.includes(id);
+      }, true)
+    );
+    setIsIndeterminate(() =>
+      childrenIds.reduce((acum, id) => {
+        return acum || value.includes(id);
+      }, false)
+    );
+  }, [value]);
```

Comment:
I think you can compute `isChecked` and `isIndeterminate` directly from the `value`, no need for the `useEffect`